### PR TITLE
Support calling __torch_function__ attribute access

### PR DESCRIPTION
**Pull request by mlazos**
Triggers `__torch_function__` tracing on attribute/method/property access matching the eager behavior for non-overridden attributes/methods/properties that are present on `torch.Tensor`.

Some caveats: 
1. for methods there doesn't seem to be a way to check if the original implementation of a method is overridden via monkey patching or not. For example:
```
class LocalSubclass(torch.Tensor):
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        return super().__torch_function__(func, types, args, kwargs)

x = torch.ones(2, 2).as_subclass(LocalSubclass)


> x.sigmoid
<built-in method sigmoid of LocalSubclass object at 0x7f8d305bb5e0>
```
There isn't a way to verify that this built-in method is equivalent to the base `torch.Tensor` implementation as each instance will have a different built-in method object that can't be traced back to the original `torch.Tensor` impl. You can check that the class itself has the original implementation via 
```
> inspect.getattr_static(LocalSubclass, "sigmoid")
<method 'sigmoid' of 'torch._C.TensorBase' objects>
```
But we can't detect if the user dynamically patches an object with a built-in method called sigmoid which does something completely different. 

2. If a user overrides a method but calls the original implementation we will still graph break. This will require modifying `SuperVariable` (and any other way to get the original impl) to handle tensor subclasses. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111765
* __->__ #111737

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_subclasses.py [hunk 1]
```diff
@@ -39,11 +66,15 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs):
         return gm
 
 
+GLOBAL_TEST_SUBCLASSES = {MockSubclass, DummyNDim, SigmoidToExpSubclass}
+
+
 @contextlib.contextmanager
 def preserve_subclass_config():
```
Comment (jansel) [inline test/dynamo/test_subclasses.py:73]: Can't this just be `torch._dynamo.config.patch`

File: torch/_dynamo/variables/torch_function.py [hunk 2]
```diff
@@ -126,6 +126,30 @@ def subclass_type_var(self):
...
+        import torch
+        from .builder import SourcelessBuilder
+
+        if name in banned_attrs:
+            unimplemented(
+                f"Accessing {name} on a tensor subclass with a __torch_function__ override is not supported"
+            )
+
+        if tx.output.torch_function_enabled:
+            get_fn = SourcelessBuilder()(tx, getattr(torch.Tensor, name).__get__)
```
Comment (jansel) [inline torch/_dynamo/variables/torch_function.py:142]: Check that `source` is not set?  Or are we sure no guards are needed?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch_function.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/torch_function.py:142]: I can create a source for this since I'm grabbing an attribute from `torch.Tensor`

Comment (jansel) [reply torch/_dynamo/variables/torch_function.py:154]: I think we need some sort of check that the user defined class doesn't overload `name`.  If the tensor subclass defines `name`, then `getattr(torch.Tensor, name)` will be wrong.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch_function.py [hunk 3]
```diff
@@ -45,6 +46,24 @@ def is_torch_function_user_object(obj):
     return hasattr(obj, "__torch_function__")
 
 
+def _is_attr_overidden(tx, var, name):
```
Comment (mlazos) [inline torch/_dynamo/variables/torch_function.py:49]: @jansel let me know what you think of this as a check

File: torch/_dynamo/variables/torch_function.py [hunk 4]
```diff
@@ -45,6 +46,24 @@ def is_torch_function_user_object(obj):
...
+    import torch
+
+    try:
+        attr_val = inspect.getattr_static(var.python_type(), name)
+        return attr_val != getattr(torch.Tensor, name)
+    except AttributeError:
+        pass
+
+    try:
+        var.dynamic_getattr(tx, name)
```
Comment (jansel) [inline torch/_dynamo/variables/torch_function.py:59]: Do we need to guard on this?

**Comment by ezyang** [comment]
needs more pr desc?

File: torch/_dynamo/variables/tensor.py [hunk 5]
```diff
@@ -1010,3 +1010,12 @@ def call_function(
...
         return super().call_function(tx, args, kwargs)
+
+    def is_python_constant(self):
+        return True
+
+    def python_type(self):
+        return type(self.value)
+
+    def as_python_constant(self):
+        return self.value
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1021]: UserDefinedObjectVariable does not report that it `is_python_constant`. Why should TensorSubclassVariable be any different?

Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:1021]: This is more analagous to UserDefinedClassVariable, which is considered a constant

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1021]: @jansel I suppose has the final word here, but this sounds like UserDefinedClassVariable is wrong to me. In general, Dynamo uses `is_python_constant` to test if something is a constant and thus Dynamo can directly manipulate it (e.g., putting these python constants into the graph, constant folding by directly working with the underlying Python value). An arbitrary defined user object never has this constant-ness; a user could have attached a metaclass to the class, there might be weird class variable mutation going on, etc.

Comment (jansel) [reply torch/_dynamo/variables/tensor.py:1021]: I tend to agree with that logic...
1) Are we guarding on ID to ensure it actually is a constant
2) Are we checking for metaclasses that could introduce side effects while compiling?
3) What code needs this is_python_constant stuff to work?

File: torch/_dynamo/variables/torch_function.py (hunk 4 above)
Comment (mlazos) [reply torch/_dynamo/variables/torch_function.py]: I removed the dynamic_getattr piece because it was giving some false positives. I added guards to ensure the an attribute/method/property is not overridden though if we make it past this check.

File: torch/_dynamo/variables/torch_function.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/torch_function.py]: I checked if a property/method/attr is overridden on the class and trigger a graph break in this case. I added guards if the source is valid to ensure that the impl of a method/property/attr on the class matches the `torch.Tensor` impl as well. Handling Ed's case will be a bit trickier with `super()` handling for subclasses so I'm going to fallback for now.

File: torch/_dynamo/variables/tensor.py (hunk 5 above)
Comment (mlazos) [reply torch/_dynamo/variables/tensor.py:1021]: I'll remove for now, at least on my local run of the subclass tests it's passing.

File: torch/_dynamo/variables/torch_function.py [hunk 6]
```diff
@@ -45,6 +46,19 @@ def is_torch_function_user_object(obj):
...
 
+def _is_attr_overidden(tx, var, name):
+    import torch
+
+    overridden = False
+    try:
+        attr_val = inspect.getattr_static(var.python_type(), name)
+        overridden |= attr_val != getattr(torch.Tensor, name)
+    except AttributeError:
+        pass
```
Comment (jansel) [inline torch/_dynamo/variables/torch_function.py:57]: When do we hit this case?   Is returning false correct?

Comment (mlazos) [reply torch/_dynamo/variables/torch_function.py:57]: Basically if a class overrides a property or method, `attr_val != getattr(torch.Tensor, name)` will be True, but if the class has the standard `torch.Tensor` implementation this will be False. I added test cases [here](https://github.com/pytorch/pytorch/pull/111737/files#:~:text=def%20test_user_overidden_attr_unsupported(self)%3A) to verify that this works.

Comment (ezyang) [reply torch/_dynamo/variables/torch_function.py:57]: Instead of catching AttributeError, you could do getattr(torch.Tensor, name, None)

File: torch/_dynamo/variables/torch_function.py [hunk 7]
```diff
@@ -160,11 +213,24 @@ def call_method(
...
+                unimplemented(
+                    f"Calling overidden method {name} on a tensor"
+                    " subclass with a __torch_function__ override is not supported"
+                )
 
             # [Note: __torch_function__] Currently we only support methods that are defined on tensor
             # we will graph break in other cases this will need a bigger overhaul of extracting methods/comparing them for equality
-            func_var = SourcelessBuilder()(tx, getattr(torch.Tensor, name))
+            # We've established with the above check that the method is not overridden, so we guard that the method is the same
+            # as the impl defined on tensor and retrieve it
```
Comment (ezyang) [inline torch/_dynamo/variables/torch_function.py:227]: Hmm... this feels like over-guarding to me. This seems to imply that we will generate a guard for EVERY function access on the subclass to guard against the user post-facto monkeypatching their class to start overriding a method. It would be much better to do one of the version dict things we were doing for nn.Module to just test that, yes, the user hasn't modified their user defined class in the meantime.

**Review by ezyang — APPROVED** [review]
There are some parts that are dorky, especially the overguarding, but ok

**Review by jansel — APPROVED** [review]
